### PR TITLE
fix the set_value error in cpu

### PR DESCRIPTION
**Pull request by zoooo0820**
### PR types

Bug fixes
### PR changes

OPs
### Describe

Pcard-66985
This PR is to fix the Bug mentioned in https://github.com/PaddlePaddle/Paddle/issues/49659 .
On CPU, when the dimension size of indexed Tensor is less than value Tensor, the result is incorrect.

Example1 (The indexed shape is [3, ] and value shape is [1, 3]):
```
>>> paddle.set_device('cpu')
Place(cpu)
>>> a = paddle.ones((3,3))
>>> a
Tensor(shape=[3, 3], dtype=float32, place=Place(cpu), stop_gradient=True,
       [[1., 1., 1.],
        [1., 1., 1.],
        [1., 1., 1.]])
>>> b = paddle.ones((1,3))*3
>>> b
Tensor(shape=[1, 3], dtype=float32, place=Place(cpu), stop_gradient=True,
       [[3., 3., 3.]])
>>> a[:, 1] = b
>>> a
Tensor(shape=[3, 3], dtype=float32, place=Place(cpu), stop_gradient=True,
       [[ 1., -3.,  1.],
        [ 1., -3.,  1.],
        [ 1., -3.,  1.]])

# all -3 should be 3
```

Example2 (The indexed shape is [2,2], and the value shape is [1, 1, 2]):
```
>>> a = paddle.zeros((2,2))
>>> a
Tensor(shape=[2, 2], dtype=float32, place=Place(cpu), stop_gradient=True,
       [[0., 0.],
        [0., 0.]])
>>> b = paddle.ones((1,1,2))
>>> b
Tensor(shape=[1, 1, 2], dtype=float32, place=Place(cpu), stop_gradient=True,
       [[[1., 1.]]])
>>> a[:] = b
>>> a
Tensor(shape=[2, 2], dtype=float32, place=Place(cpu), stop_gradient=True,
       [[-1., -1.],
        [-1., -1.]])

# all -1 should be 1
```

### CODE REVIEW

**Review by zhiqiu — COMMENTED** [review]
LGTM

File: paddle/phi/kernels/impl/set_value_kernel_impl.h [hunk 1]
```diff
@@ -185,16 +184,32 @@ void SetValueImpl(const Context& dev_ctx,
...
-      funcs::SubtractFunctor<T>(),
-      &slice_tensor);
 
+  bool is_gpu_place = dev_ctx.GetPlace().GetType() == phi::AllocationType::GPU;
+  if (is_gpu_place || slice_tensor.dims().size() >= value.dims().size()) {
+    // ElementwiseComputeEx can do broadcasting in two cases:
+    // 1. The place is GPU.
+    // 2. The place is CPU, and the 'x' does not need broadcast.
+    // Please see the note in
+    // paddle/fluid/operators/elementwise/elementwise_op_function.h
```
Comment (jeff41404) [inline paddle/phi/kernels/impl/set_value_kernel_impl.h]: I understand that this is only a workaround solution, so need to add TODO to explain the futrue work.

Comment (zoooo0820) [reply paddle/phi/kernels/impl/set_value_kernel_impl.h]: Done, thanks

**Review by jeff41404 — APPROVED** [review]
LGTM

**Pull request merged**